### PR TITLE
Upgrade kudu client to 1.15.0

### PR DESCRIPTION
**Pull request by grantatspothero**
Upgrading the kudu client revealed a few problems:
1. Timeouts to kudu tablets were sometimes occurring during deletes due to this change introduced in the kudu java client in version 1.13.0: https://github.com/apache/kudu/commit/d23ee5d38ddc4317f431dd65df0c825c00cc968a#diff-f1f50409d81052b8f8d7aea7da663c185c704c6206cb0ec901114f4d9ee8c28f
(see here for the reason why that commit broke the client: https://gerrit.cloudera.org/#/c/18166/)

2. Timeouts were *not* failing query execution because the kudu connector was configured to flush operations in the background.

3. The two combined above meant tests that did deletes sometimes actually did not perform deletes and would fail.

The first commit upgrades the kudu client and additionally explicitly fails trino execution when kudu rpcs timeout. This resolves: https://github.com/trinodb/trino/issues/5687

The second commit fixes the problem of one source of kudu timeouts in the 1.15.0 client .

### CODE REVIEW

**Comment by grantatspothero** [comment]
Want to let the CI process run so there is a build showing the timeouts causing query failure.

After that, will push another commit to actually resolve the timeout problems by setting `KuduScanToken.KuduScanTokenBuilder.includeTabletMetadata(false)`

**Comment by grantatspothero** [comment]
Also if we want to split the kudu client upgrade from the background flushing problem let me know. 

There just is not a way to confirm the background flushing problem with kudu client 1.10.0 since the timeouts never/rarely happen.

File: plugin/trino-kudu/src/main/java/io/trino/plugin/kudu/TypeHelper.java
Comment (grantatspothero) [inline plugin/trino-kudu/src/main/java/io/trino/plugin/kudu/TypeHelper.java:126]: Since this PR is already complex enough, figure adding support for the new types (varchar/date) could come at a later time.

File: plugin/trino-kudu/src/main/java/io/trino/plugin/kudu/KuduPageSink.java
Comment (Praveen2112) [inline plugin/trino-kudu/src/main/java/io/trino/plugin/kudu/KuduPageSink.java:109]: Was this introduced in the latest version ?

File: plugin/trino-kudu/src/main/java/io/trino/plugin/kudu/KuduPageSink.java
Comment (Praveen2112) [inline plugin/trino-kudu/src/main/java/io/trino/plugin/kudu/KuduPageSink.java:131]: Can we extract it from the version update ? It looks like `OperationResponse` is returned in old APIs too.

File: plugin/trino-kudu/src/main/java/io/trino/plugin/kudu/KuduRecordCursor.java
Comment (Praveen2112) [inline plugin/trino-kudu/src/main/java/io/trino/plugin/kudu/KuduRecordCursor.java:125]: We could also extract this as a separate commit right ?

File: plugin/trino-kudu/src/test/java/io/trino/plugin/kudu/TestingKuduServer.java
Comment (Praveen2112) [inline plugin/trino-kudu/src/test/java/io/trino/plugin/kudu/TestingKuduServer.java:37]: Can it be a part of the first commit ?

File: plugin/trino-kudu/src/main/java/io/trino/plugin/kudu/KuduPageSink.java
Comment (grantatspothero) [reply plugin/trino-kudu/src/main/java/io/trino/plugin/kudu/KuduPageSink.java:109]: No, this config also existed in 1.10.0. 

See the PR description, the only way I could reproduce this bug (https://github.com/trinodb/trino/issues/5687) was by upgrading the kudu client to 1.15.0 which triggered timeouts to kudu. 

That is why I fixed this bug in this upgrade client PR.

File: plugin/trino-kudu/src/main/java/io/trino/plugin/kudu/KuduPageSink.java
Comment (grantatspothero) [reply plugin/trino-kudu/src/main/java/io/trino/plugin/kudu/KuduPageSink.java:131]: Same comment as above, there is no way to easily verify the fix because you need a very specific sequence of timeouts [no timeouts when initially connecting to kudu, but timeouts during deletes/upserts/etc]. The kudu 1.15.0 client [accidentally] provides those timeouts due to a bug in the client.

See here for the failing kudu tests when upgrading to kudu 1.15.0:
https://github.com/trinodb/trino/runs/5057440792?check_suite_focus=true

And here for the workaround that makes tests pass:
https://github.com/trinodb/trino/runs/5060298096?check_suite_focus=true

If we do not mind having no tests around the change, I can extract this into a separate PR.

File: plugin/trino-kudu/src/main/java/io/trino/plugin/kudu/KuduRecordCursor.java
Comment (grantatspothero) [reply plugin/trino-kudu/src/main/java/io/trino/plugin/kudu/KuduRecordCursor.java:125]: Yes, will extract this into a separate commit

File: plugin/trino-kudu/src/test/java/io/trino/plugin/kudu/TestingKuduServer.java
Comment (grantatspothero) [reply plugin/trino-kudu/src/test/java/io/trino/plugin/kudu/TestingKuduServer.java:37]: Sure, I wanted to verify the kudu 1.15 client could also talk to kudu 1.10 server. The docs say the wire protocol is backwards compatible, but wanted to verify.

But I can move this into one commit now.

File: plugin/trino-kudu/src/main/java/io/trino/plugin/kudu/KuduRecordCursor.java
Comment (grantatspothero) [reply plugin/trino-kudu/src/main/java/io/trino/plugin/kudu/KuduRecordCursor.java:125]: Separate PR is here: https://github.com/trinodb/trino/pull/10986

File: plugin/trino-kudu/src/main/java/io/trino/plugin/kudu/TypeHelper.java
Comment (hashhar) [reply plugin/trino-kudu/src/main/java/io/trino/plugin/kudu/TypeHelper.java:126]: Can you please create an issue and link to it here in the todo? (and below)

**Review by hashhar — COMMENTED** [review]
I believe all needs to be squashed except the last commit.

LGTM

File: plugin/trino-kudu/src/main/java/io/trino/plugin/kudu/TypeHelper.java
Comment (grantatspothero) [reply plugin/trino-kudu/src/main/java/io/trino/plugin/kudu/TypeHelper.java:126]: Issue here: https://github.com/trinodb/trino/issues/11009

**Review by hashhar — APPROVED** [review]
LGTM % doc update.

Please update "Requirements" in Kudu docs to specify minimum supported version as 1.15 now. It might work with older versions but we don't test it now - if we intend to claim otherwise then add a test extending BaseConnectorSmokeTest with an older Kudu version.

**Comment by grantatspothero** [comment]
> LGTM % doc update.
> 
> Please update "Requirements" in Kudu docs to specify minimum supported version as 1.15 now. It might work with older versions but we don't test it now - if we intend to claim otherwise then add a test extending BaseConnectorSmokeTest with an older Kudu version.

Kudu 1.10.0 is pretty old (>2years,  released on November 1, 2019) so I just updated the docs to say we only support 1.15.0 or higher

**Comment by hashhar** [comment]
That's fair. 1.13 is oldest supported release anyway according to https://kudu.apache.org/releases.

**Comment by grantatspothero** [comment]
> That's fair. 1.13 is oldest supported release anyway according to https://kudu.apache.org/releases.

If we care about supporting 1.13 lmk, I can add an additional set of smoke tests in.

**Comment by hashhar** [comment]
> > That's fair. 1.13 is oldest supported release anyway according to https://kudu.apache.org/releases.
> 
> If we care about supporting 1.13 lmk, I can add an additional set of smoke tests in.

Would be nice to do if it's not too much work otherwise we can tackle that separately. Please remember to update docs accordingly.

File: plugin/trino-kudu/src/test/java/io/trino/plugin/kudu/TestKudu115SmokeTests.java
Comment (hashhar) [inline plugin/trino-kudu/src/test/java/io/trino/plugin/kudu/TestKudu115SmokeTests.java:16]: Instead of version numbers in name we can use KuduLatestSmokeTests and KuduSmokeTests (existing convention for other connectors) since we'll never test with more than 2 versions (most likely).

File: plugin/trino-kudu/src/test/java/io/trino/plugin/kudu/TestKudu113SmokeTests.java
Comment (hashhar) [inline plugin/trino-kudu/src/test/java/io/trino/plugin/kudu/TestKudu113SmokeTests.java:20]: I'm surprised this kind of composition of tests actually works.

I wouldn't mind these being separate test classes but I'll defer to @Praveen2112 / @ebyhr.

File: plugin/trino-kudu/src/test/java/io/trino/plugin/kudu/TestingKuduServer.java
Comment (hashhar) [inline plugin/trino-kudu/src/test/java/io/trino/plugin/kudu/TestingKuduServer.java:52]: I don't think this comment is true? We can obviously mix older clients with newer servers or vice versa for most of the versions.

File: docs/src/main/sphinx/connector/kudu.rst
Comment (hashhar) [inline docs/src/main/sphinx/connector/kudu.rst:14]: In " Upgrade kudu client to 1.15.0 " please wrap the commit message body to 72 chars. See https://chris.beams.io/posts/git-commit/ for project's convention.

File: plugin/trino-kudu/src/main/java/io/trino/plugin/kudu/KuduClientConfig.java
Comment (hashhar) [inline plugin/trino-kudu/src/main/java/io/trino/plugin/kudu/KuduClientConfig.java:36]: reword commit message from `Deprecate` to `Remove` - since deprecation doesn't cause failures for end users and doesn't require changes but this will require changes.

**Review by hashhar — APPROVED** [review]
LGTM % comments.

I believe https://github.com/trinodb/trino/pull/10953/commits/8766971f78d7c8dcf0a1e101b523787e3d7b8800 depends on this PR getting merged first?

File: plugin/trino-kudu/src/main/java/io/trino/plugin/kudu/KuduPageSink.java
Comment (wendigo) [inline plugin/trino-kudu/src/main/java/io/trino/plugin/kudu/KuduPageSink.java:130]: can we extract this verification to a common static method?

`void verifyOperationSucceeded(@Nullable OperationResponse) throw TrinoException`

File: plugin/trino-kudu/src/main/java/io/trino/plugin/kudu/KuduUpdatablePageSource.java
Comment (wendigo) [inline plugin/trino-kudu/src/main/java/io/trino/plugin/kudu/KuduUpdatablePageSource.java:66]: use extracted verify method

File: plugin/trino-kudu/src/main/java/io/trino/plugin/kudu/schema/SchemaEmulationByTableNameConvention.java
Comment (wendigo) [inline plugin/trino-kudu/src/main/java/io/trino/plugin/kudu/schema/SchemaEmulationByTableNameConvention.java:73]: use extracted verify method

File: plugin/trino-kudu/src/test/java/io/trino/plugin/kudu/TestKudu113SmokeTests.java
Comment (wendigo) [reply plugin/trino-kudu/src/test/java/io/trino/plugin/kudu/TestKudu113SmokeTests.java:20]: It looks neat for me. If there are no downsides we could keep it.

Comment (grantatspothero) [reply plugin/trino-kudu/src/test/java/io/trino/plugin/kudu/TestKudu113SmokeTests.java:20]: I was surprised it worked too, but it is nice to encapsulate all the version-specific tests in one place.

File: plugin/trino-kudu/src/test/java/io/trino/plugin/kudu/TestingKuduServer.java
Comment (grantatspothero) [reply plugin/trino-kudu/src/test/java/io/trino/plugin/kudu/TestingKuduServer.java:52]: I put this comment because other tests use the default `TestingKuduServer()` constructor. Only the smoke tests specifically specify the version of the kudu server to use. 

When we upgrade the client version, we presumably want to also upgrade all the [non-smoke tests] to use a newer version of the kudu server.

File: docs/src/main/sphinx/connector/kudu.rst
Comment (grantatspothero) [reply docs/src/main/sphinx/connector/kudu.rst:14]: 👍 
Any tips for helping format commit messages? Without a linter/some automated tool in place it is something I always forget.

Comment (grantatspothero) [reply docs/src/main/sphinx/connector/kudu.rst:14]: Nevermind, was able to configure vim to auto-wrap at 72 chars for only git commit messages. 

~/.vimrc
```
"enable filetype detection
filetype plugin indent on

"auto wrap commit messages at 72 chars
autocmd FileType gitcommit set textwidth=72
autocmd FileType gitcommit set colorcolumn=73
```

**Review by hashhar — APPROVED** [review]
LGTM

File: plugin/trino-kudu/src/test/java/io/trino/plugin/kudu/KuduSmokeTests.java
Comment (findepi) [inline plugin/trino-kudu/src/test/java/io/trino/plugin/kudu/KuduSmokeTests.java:20]: Any reason this is not defined as a top-level class?

Comment (hashhar) [reply plugin/trino-kudu/src/test/java/io/trino/plugin/kudu/KuduSmokeTests.java:20]: See https://github.com/trinodb/trino/pull/10940#discussion_r806479586. I believe to keep all tests for same version together.

Comment (findepi) [reply plugin/trino-kudu/src/test/java/io/trino/plugin/kudu/KuduSmokeTests.java:20]: Sharing a constant doesn't require nesting classes.
i admire cleverness, but this also means unnecessary class hierarchy, which doesn't help browse the code.

if the paradigm was more frequent in the code base, i would probably get used to it and wouldn't complain.

Comment (grantatspothero) [reply plugin/trino-kudu/src/test/java/io/trino/plugin/kudu/KuduSmokeTests.java:20]: I noticed depending on how you run the test `mvn test -Dtest=io.trino.plugin.kudu.KuduLatestConnectorTests` vs `mvn test` some tests will be skipped in the class hierarchy. 

Additionally, when trying to run tests through intellij's UI you can only run the static inner classes (not the top level class).

Seems like the tooling support for static inner test classes is just not good, I'm going to move these to top level classes

**Pull request merged**
